### PR TITLE
V2 of the action

### LINKED ISSUE #8: Use pre-built binaries to make action faster
This action is our slowest PR validation check. I routinely see this action take ~2 minutes to build cargo-semver-checks from scratch on each PR validation run.

Please consider releasing binaries as part of your crate release process and downloading those boundaries in this GitHub action.

I refer you to https://github.com/aig787/cargo-udeps-action, which does this and is on-task in just a few seconds.

(Yes, I recognize the building-docs steps are going to be slow regardless.)

### LINKED ISSUE #16: Allow enabling verbose output
To provide increased insight into which checks were run even when successful, I would like to enable verbose output. It does not seem like the action provides a way to do that presently.

### LINKED ISSUE #18: Add `manifest-path` option to specify the path to the Cargo.toml to check
Hi,

As far as I can tell, the action currently assumes there's a `Cargo.toml` at the root of the repository (either a crate or a workspace).

Could it be possible to specify the directory containing the `Cargo.toml`? (the default would be `.` i.e. the root of the repository)

Thanks!

### LINKED ISSUE #23: Different behavior between the Github action and the CLI
### Steps to reproduce the bug with the above code

I was upgrading a crate from `0.1.0-rc2` to `0.1.0`. No actual code changes were present, just exiting the release candidate phase.

### Actual Behaviour

I used

```yml
      - name: Check semver
        uses: obi1kenobi/cargo-semver-checks-action@v1
        with:
          crate-name: qr-rs-lib
```

as part of my CD pipeline and got [this](https://github.com/AntoniosBarotsis/qr-rs/actions/runs/4197182627/jobs/7279161633) error:

```
Crate qr-rs-lib current version: 0.1.0
Versions on crates.io:
0.1.0-rc2
0.1.0-rc1

There is no suitable comparison version.
The current version 0.1.0 is smaller than any version published on crates.io
+ export COMPARISON_TAG=v
+ COMPARISON_TAG=v
+ git fetch --depth=1 origin +refs/tags/v:refs/tags/v
fatal: couldn't find remote ref refs/tags/v
Error: Process completed with exit code 128.
```

My workflow can be found [here](https://github.com/AntoniosBarotsis/qr-rs/blob/8d2412ebdba0c51236a67d188dc9df02c4793adc/.github/workflows/release-lib.yml) and the relevant part is [this](https://github.com/AntoniosBarotsis/qr-rs/blob/8d2412ebdba0c51236a67d188dc9df02c4793adc/.github/workflows/release-lib.yml#L17-L20).

### Expected Behaviour

This should not throw an error. The interesting thing is how the CLI does not consider this to be wrong (which is the expected behavior).

In fact, [that is how I *fixed* my pipeline, by using the CLI instead of the action](https://github.com/AntoniosBarotsis/qr-rs/commit/d66622f0400857d660fa6c9a3ae0f067aa797795)

```yml
      - name: Check semver
        working-directory: lib
        run: |
          cargo install cargo-semver-checks --locked
          cargo semver-checks check-release
```

#### Software version

cargo-semver-checks 0.17.1 (af71be0)

#### Operating system

Windows 6.2.9200

#### Command-line

```bash
C:\Users\anton\.cargo\bin\cargo-semver-checks.exe semver-checks --bugreport
```

#### cargo version

```
> cargo -V
cargo 1.67.1 (8ecd4f20a 2023-01-10)
```

#### Compile time information

- Profile: release
- Target triple: x86_64-pc-windows-msvc
- Family: windows
- OS: windows
- Architecture: x86_64
- Pointer width: 64
- Endian: little
- CPU features: fxsr,sse,sse2
- Host: x86_64-pc-windows-msvc

### Build Configuration

My `config.toml` can be found [here](https://github.com/AntoniosBarotsis/qr-rs/blob/8d2412ebdba0c51236a67d188dc9df02c4793adc/.cargo/config.toml) however, it only contains *~way too many~* lints and was present both in the failing and passing scenarios.

### Additional Context

_No response_

### PR DESCRIPTION
**Pull request by mgr0dzicki**
Uses `cargo-semver-checks` built-in logic instead of the bash scripts (closes #23).
Introduces option `manifest-path` (closes #18).
Introduces option `verbose` (closes #16).
Uses pre-built binaries (closes #8).

Because the default behavior is changed and inputs `crate-target` and `version-tag-prefix` are removed, a version bump to v2 is necessary.

### CODE REVIEW

**Comment by mgr0dzicki** [comment]
While testing the draft of new action on GitHub `windows-latest` runner, I ran into a following error:

```
C:\Users\runneradmin\.cargo\bin\cargo.exe semver-checks check-release
Error: failed to make directory 'C:\Users\runneradmin\.cargo\registry/index\github.com-1ecc6299db9ec823': The system cannot find the path specified.
; class=Os (2)

Caused by:
    failed to make directory 'C:\Users\runneradmin\.cargo\registry/index\github.com-1ecc6299db9ec823': The system cannot find the path specified.
    ; class=Os (2)
Error: The process 'C:\Users\runneradmin\.cargo\bin\cargo.exe' failed with exit code 1
```

@obi1kenobi do you think it is a problem with the setup done by the action? Or might it be an issue with the `cargo-semver-checks` itself? I don't have a working Windows installation, so I didn't check it locally yet.

**Comment by obi1kenobi** [comment]
> Or might it be an issue with the cargo-semver-checks itself?

This looks like a bug in `cargo-semver-checks` itself: it looks like it isn't properly constructing subdirectory names and using the wrong directory separator (the Unix `/` instead of the Windows `\`). Nice catch! Would you care to reproduce the problem in the cargo-semver-checks repo itself, and then debug and resolve it?

File: src/main.ts [hunk 1]
```diff
@@ -0,0 +1,124 @@
...
+async function getCargoSemverChecksDownloadURL(target: string): Promise<string> {
+    const octokit = github.getOctokit(getGitHubToken());
+
+    const getReleaseUrl = await octokit.rest.repos.getLatestRelease({
+        owner: 'obi1kenobi',
+        repo: 'cargo-semver-checks'
+    });
+
+    const asset = getReleaseUrl.data.assets.find(asset => {
+        return asset['name'].endsWith(`${target}.tar.gz`)
```
Comment (obi1kenobi) [inline src/main.ts:51]: Is it possible the release isn't always a `.tar.gz`? Windows in particular seems to be a `.zip`.

File: src/main.ts [hunk 2]
```diff
@@ -0,0 +1,124 @@
...
+
+async function installCargoSemverChecksFromPrecompiledBinary(): Promise<void> {
+    const url = await getCargoSemverChecksDownloadURL(getPlatformMatchingTarget());
+
+    core.info(`downloading cargo-semver-checks from ${url}`);
+    const tarballPath = await toolCache.downloadTool(url, undefined, `token ${getGitHubToken()}`, {
+        accept: 'application/octet-stream'
+    });
+    core.info(`extracting ${tarballPath}`);
+    const binPath = await toolCache.extractTar(tarballPath, undefined, ["xz"]);
```
Comment (obi1kenobi) [inline src/main.ts:80]: If the file isn't a tar (e.g. on Windows), this looks like it would fail?

File: src/main.ts [hunk 3]
```diff
@@ -0,0 +1,124 @@
...
+
+async function main() {
+    try {
+        await run();
+    } catch (error: any) {
+        core.setFailed(error.message);
+    }
+}
+
+main();
```
Comment (obi1kenobi) [inline src/main.ts:124]:
```suggestion
main();

```

Comment (obi1kenobi) [reply src/main.ts:124]: Also, should this `await` the call? I'm not up to date on my TypeScript semantics here.

File: tsconfig.json [hunk 4]
```diff
@@ -0,0 +1,10 @@
+{
+    "compilerOptions": {
+        "module": "commonjs",
+        "outDir": "dist",
+        "strict": true,
+    },
+    "include": [
+        "src/**/*.ts"
+    ]
+}
```
Comment (obi1kenobi) [inline tsconfig.json:10]:
```suggestion
}

```

File: .github/workflows/ci.yml [hunk 5]
```diff
@@ -6,52 +6,21 @@ on:
     branches:
       - main
 
-env:
-  RUST_BACKTRACE: 1
-  CARGO_TERM_COLOR: always
-
 jobs:
```
Comment (obi1kenobi) [inline .github/workflows/ci.yml:9]: Perhaps add a test that the TypeScript source code matches the JavaScript in the `dist` directory? Otherwise we might get into a bad situation where we think something is implemented but we've forgotten to recompile and it isn't actually running.

**Review by obi1kenobi — COMMENTED** [review]
Excited to see this coming together nicely! 🎉

File: src/main.ts [hunk 6]
```diff
@@ -0,0 +1,124 @@
...
+
+function getPlatformMatchingTarget(): string {
+    const platform = os.platform() as string;
+    switch (platform) {
+        case "linux":
+            return "x86_64-unknown-linux-gnu"
+        case "win32":
+            return "x86_64-pc-windows-msvc"
+        case "darwin":
+            return "x86_64-apple-darwin"
```
[the remark this answers is not in the document]
Comment (AntoniosBarotsis) [reply src/main.ts:17]: I don't think GA supports that [yet](https://github.com/actions/runner-images/issues/5631)?

File: action.yml [hunk 7]
```diff
@@ -5,76 +5,18 @@ branding:
...
-        # Build rustdoc JSON for the current version, and move it to /tmp/
-        # so it doesn't get overwritten by the baseline build.
-        RUSTC_BOOTSTRAP=1 cargo rustdoc $RUSTDOC_EARLY_FLAGS -- $RUSTDOC_LATE_FLAGS
-        mv "$(cargo metadata --format-version 1 | jq -r .target_directory)/doc/$PACKAGE_NAME_WITH_UNDERSCORES.json" /tmp/current.json
-
-        # Check for semver violations.
-        cargo install cargo-semver-checks --locked
-        cargo semver-checks check-release --current /tmp/current.json --baseline /tmp/baseline.json
+  using: 'node16'
+  main: 'dist/index.js'
```
[the remark this answers is not in the document]
Comment (mgr0dzicki) [reply action.yml]: I should have done it long time ago, sorry for that, fixed!

File: src/main.ts (hunk 3 above)
Comment (mgr0dzicki) [reply src/main.ts]: Great question! I had to check in order to make myself sure. The promise is executed even if it is not awaited. And in fact, at the top level of a script the `await` is even not allowed.

File: .github/workflows/ci.yml (hunk 5 above)
Comment (mgr0dzicki) [reply .github/workflows/ci.yml:9]: Great idea! I've already made this mistake once or twice...

File: src/main.ts (hunk 1 above)
Comment (mgr0dzicki) [reply src/main.ts:51]: There is also a release for Windows packed as `.tar.gz` and it seems to be exactly the same binary as the one in the `.zip` file, am I right? If yes, then it is easier to always download the `.tar.gz` archive as it doesn't require separating the cases.

File: README.md [hunk 8]
```diff
@@ -19,55 +14,24 @@ Every argument is optional.
...
-- name: Publish my-crate to crates.io
-  run: # your `cargo publish` code here
-```
 
 ## Use in workspaces with more than one crate
 
-You'll need to specify which crate should be checked, and the format used for version tags for that crate:
-- `crate-name` specifies the crate to check, and
-- `version-tag-prefix` sets the text prepended to the version number to create the git tag for that release.
+By default, if workspace contains multiple crates, all of them are checked against semver violations. You can specify single crate to be checked instead using `crate-name` or `manifest-path`.
```
[the remark this answers is not in the document]
Comment (mgr0dzicki) [reply README.md]: It might be useful indeed, I will add it!

File: src/main.ts (hunk 6 above)
[the remark this answers is not in the document]
Comment (mgr0dzicki) [reply src/main.ts:17]: As far as I know, all runners hosted by GitHub use x86_64 architecture. We might consider adding support for other platforms, but the questions are:

1. Does anyone really need this action running in such a specific environment like the self-hosted runner based on for example arm macOS?
2. How would we test the support for other platforms? It would require using a runner different from these provided by GitHub. I don't have access to any arm-based device at the moment (other then my phone :-p), so I'm not sure if I'm able to do it.

File: .github/workflows/ci.yml (hunk 5 above)
Comment (mgr0dzicki) [reply .github/workflows/ci.yml:9]: Aaaand done! It would be great if you could take a look :)

**Comment by mgr0dzicki** [comment]
I've found another (and the last) issue causing the tool to fail on Windows - the directory `.cargo\registry\index` does not exist at the beginning and therefore the creation of its subdirectory fails. The current workaround is https://github.com/obi1kenobi/cargo-semver-checks-action/pull/21/commits/851a833b93d27886a6062077ea065773c6583038.

@obi1kenobi, I've seen that you were successful in debugging the previous issue and fixing it in https://github.com/frewsxcv/rust-crates-index/pull/94. Could you have a look at this problem too? I believe it shouldn't take you much time, as it is probably another simple fix in `rust-crates-index` code, which you've already worked with. Of course if you don't have enough time then I will take it ;) However I think I would first need to setup a virtual machine with Windows to debug it locally...

BTW, I'm back to work and hope to get v2 of the action finished soon!

File: src/main.ts (hunk 6 above)
[the remark this answers is not in the document]
Comment (obi1kenobi) [reply src/main.ts:18]: If GitHub Actions doesn't support them then we don't have to add them now either. Let's make an issue to revisit this in the future and link to the upstream issue for ARM-based macOS and Linux builds.

In general, semver-checking is sensitive to the platform because conditional compilation allows [parameterizing based on the platform](https://doc.rust-lang.org/reference/conditional-compilation.html#target_arch). The generated rustdoc JSON represents the configured platform. Most crates don't have platform-specific differences in API, but some lower-level crates probably do and we ideally would support them: if semver-checking is hard on one platform, it's even harder when you might semver-break only "x86 on linux with avx2 enabled."

File: src/main.ts (hunk 1 above)
Comment (obi1kenobi) [reply src/main.ts:52]: Good call, I didn't notice that release.

File: src/main.ts (hunk 3 above)
Comment (obi1kenobi) [reply src/main.ts]: Ugh. I can't believe top-level `await` is still not a supported feature, it's been in the works for ages.

**Comment by obi1kenobi** [comment]
> I've found another (and the last) issue causing the tool to fail on Windows - the directory `.cargo\registry\index` does not exist at the beginning and therefore the creation of its subdirectory fails. The current workaround is [851a833](https://github.com/obi1kenobi/cargo-semver-checks-action/commit/851a833b93d27886a6062077ea065773c6583038).

This seems like it probably means `cargo-semver-checks` on Windows itself is broken too. Would you mind extracting a test case from this and adding it to the integration suite for `cargo-semver-checks`? I don't think you'd need a Windows VM for this.

Once we have that minimal test case, it will be easy to figure out whether it's something that `cargo-semver-checks` does incorrectly or whether we need an upstream fix in one of our dependencies, and if you'd prefer then I can take over from there.

File: src/main.ts [hunk 9]
```diff
@@ -0,0 +1,129 @@
...
+        optionIfValueProvided('--package', rustCore.input.getInput('crate-name')),
+        optionIfValueProvided('--manifest-path', rustCore.input.getInput('manifest-path')),
+        rustCore.input.getInputBool('verbose') ? ' --verbose' : ''
+    ].filter(el => el != '');
+}
+
+function getGitHubToken(): string {
+    const token = process.env['GITHUB_TOKEN'] || rustCore.input.getInput('github-token');
+    if (!token)
+        throw new Error('Querying the GitHub API is possible only if the GitHub token is set.');
```
Comment (obi1kenobi) [inline src/main.ts:39]: Let's try to always use curly braces with statements like `if / for / while` etc.

Otherwise, we risk having [a `goto fail;` style bug](https://nakedsecurity.sophos.com/2014/02/24/anatomy-of-a-goto-fail-apples-ssl-bug-explained-plus-an-unofficial-patch/) which would be really annoying to track down and deal with. The 1-character savings when not putting in the curly braces is extremely not worth the risk.

File: .github/workflows/ci.yml (hunk 5 above)
Comment (obi1kenobi) [reply .github/workflows/ci.yml:9]: Neat! I trust you've already tested that this "fails correctly" and catches the mistake we intended it to catch?

File: .github/workflows/ci.yml [hunk 10]
```diff
@@ -6,52 +6,25 @@ on:
...
+  compare-dist:
+    name: Check if compiled action in dist/ matches the source files
+    uses: ./.github/workflows/compare-dist.yml
+
```
Comment (obi1kenobi) [inline .github/workflows/ci.yml:13]: Can we add a TypeScript linter pass as well? Linters and auto-formatters are best added as early as possible — less lint we'll have to fix down the line.

I'm not an expert on the subject but the advice in these links looks credible and reasonably up-to-date:
https://blog.logrocket.com/linting-typescript-eslint-prettier/
https://blog.logrocket.com/using-prettier-eslint-automate-formatting-fixing-javascript/

File: src/main.ts [hunk 11]
```diff
@@ -0,0 +1,129 @@
...
+    }
+}
+
+function optionIfValueProvided(option: string, value?: string): string {
+    return value ? ` ${option} ${value}` : '';
+}
+
+function getCheckReleaseArguments(): string[] {
+    return [
+        optionIfValueProvided('--package', rustCore.input.getInput('crate-name')),
```
Comment (obi1kenobi) [inline src/main.ts:30]: Since we're already making breaking changes to the action, this is a great time to change `crate-name` to `package`. That makes it match the underlying option name, and is also cleaner since the package spec is not always just a crate name: https://doc.rust-lang.org/cargo/reference/pkgid-spec.html

Ideally, we'd link to the above in the README when explaining what the `package` does.

**Review by obi1kenobi — COMMENTED** [review]
Definitely moving in the right direction! Just a few "setting ourselves up for future success" suggestions.

File: .github/workflows/ci.yml (hunk 5 above)
Comment (mgr0dzicki) [reply .github/workflows/ci.yml:9]: Sure! I'm developing things like these on my private fork to make this pull request clean. This is a draft PR: https://github.com/mgr0dzicki/cargo-semver-checks-action/pull/8, and especially the "correctly failing" workflow: https://github.com/mgr0dzicki/cargo-semver-checks-action/actions/runs/4232011758/jobs/7351233661.

File: src/main.ts (hunk 9 above)
Comment (mgr0dzicki) [reply src/main.ts]: Changed! I've also added the ESLint rule for this.

File: .github/workflows/ci.yml (hunk 10 above)
Comment (mgr0dzicki) [reply .github/workflows/ci.yml:13]: You're right, these are definitely must-haves! I've added both ESLint and Prettier checks.

**Comment by mgr0dzicki** [comment]
> This seems like it probably means `cargo-semver-checks` on Windows itself is broken too. Would you mind extracting a test case from this and adding it to the integration suite for `cargo-semver-checks`? I don't think you'd need a Windows VM for this.

Sure, I'm currently working on that :).

File: src/main.ts (hunk 6 above)
[the remark this answers is not in the document]
Comment (mgr0dzicki) [reply src/main.ts]: Does the upstream cargo-semver-checks issue regarding ARM-based systems already exists? I can't find any.

[the remark this answers is not in the document]
Comment (obi1kenobi) [reply src/main.ts]: Feel free to open one and add the context from here.

It isn't clear what cargo-semver-checks could do in this case, other than document the limitation and advise users to generate rustdoc (or run the entire checking process) on the target they care about.

On the cargo-semver-checks side, a bigger limitation right now is the inability to specify features to check, so the target-specific issue has taken the back seat at the moment.

[the remark this answers is not in the document]
Comment (mgr0dzicki) [reply src/main.ts]: Added the issue for future: https://github.com/obi1kenobi/cargo-semver-checks-action/issues/24.

File: .github/workflows/test-build.yml [hunk 12]
```diff
@@ -0,0 +1,52 @@
...
+      - name: Setup Node.js 16.x
+        uses: actions/setup-node@v3
+        with:
+          node-version: 16
+      - name: Install npm-cli-login
+        run: npm install -g npm-cli-login
+      - name: Login to GitHub npm registry
+        run: |
+          npm-cli-login -s @actions-rs -r https://npm.pkg.github.com \
+            -u $ -p ${{ secrets.GITHUB_TOKEN }} -e test@example.com
```
Comment (obi1kenobi) [inline .github/workflows/test-build.yml:22]: Is the `-u $` intended here? It seems like a strange choice of syntax since the `$` is a special character both for GitHub Actions and for the shell 🤔

File: README.md [hunk 13]
```diff
@@ -19,55 +14,25 @@ Every argument is optional.
...
 |--------------------|-----------------------------------------------------------------------------------------------------------------------------------|---------|
-| crate-name         | The crate whose API to check for semver (needed only in multi-crate workspaces). | |
-| crate-target       | (Deprecated) By default, check the library target of the crate. To check a different target (e.g. a binary target), set this to `--bin <NAME>`. Will be removed in future versions, as bin targets do not have library-like API to check. | `--lib` |
-| version-tag-prefix | The prefix to use for the git tag for a version; the default "v" creates tags like "v1.0.0". | `v` |
+| github-token       | The `GITHUB_TOKEN` secret, which is necessary to download precompiled binaries from GitHub API. By default, the [automatic GitHub token](https://docs.github.com/en/actions/security-guides/automatic-token-authentication) provided to the workflow will be used. The token may be alternatively passed in an environment variable `GITHUB_TOKEN`. | |
```
Comment (obi1kenobi) [inline README.md:17]: Is this required? It feels like it should be optional but it never really clearly says so anywhere. If I understand this correctly, this is an optional override, not a required input, so let's frame the description like that.

File: README.md [hunk 14]
```diff
@@ -19,55 +14,25 @@ Every argument is optional.
...
+| package            | The package whose API to check for semver (in Package Id Specification format, see https://doc.rust-lang.org/cargo/reference/pkgid-spec.html for reference). If not set, all packages in the workspace are processed. | |
+| manifest-path      | Path to Cargo.toml of crate or workspace to check. Has an effect only if `package` is not specified. | |
+| verbose            | Enables verbose output of `cargo-semver-checks`. | `false` |
```
Comment (obi1kenobi) [inline README.md:20]: Let's think through the order in which we're presenting these options. They should be ordered in "always-needed -> rarely-needed" order

**Comment by obi1kenobi** [comment]
Needs a bit of polish on the README, otherwise looks in pretty good shape!

What else does it need before it's ready to merge? I know we wanted to add baseline rustdoc caching, but we can do that in a separate PR and it'll be easier to review that way.

**Comment by mgr0dzicki** [comment]
> Needs a bit of polish on the README, otherwise looks in pretty good shape!
> 
> What else does it need before it's ready to merge? I know we wanted to add baseline rustdoc caching, but we can do that in a separate PR and it'll be easier to review that way.

Thank you for all the comments and remarks, they are really valuable! I think the action is ready itself. I still want to add some workflow testing the action inputs (none of them is used at the moment). And then I think that polishing README and adding caching will be the only things remaining.

**Comment by obi1kenobi** [comment]
Sounds great! Thanks for all the awesome work you've been putting in!

File: .github/workflows/test-build.yml (hunk 12 above)
Comment (mgr0dzicki) [reply .github/workflows/test-build.yml]: You're right, it is definitely not the best placeholder... I changed it a little and added a comment, please have a look now ;)

**Comment by mgr0dzicki** [comment]
I polished the README, please take a look and let me know what do you think now :)

In the meanwhile, I'll start to work on rustdoc caching in a separate PR and notify you once it is ready for review.

File: .github/workflows/test-action.yml [hunk 15]
```diff
@@ -0,0 +1,50 @@
+name: Smoke test the action
```
Comment (obi1kenobi) [inline .github/workflows/test-action.yml:1]: Consider adding substantially more info here, maybe even a README in the directory?

The testing setup is non-trivial, and it's hard to understand which piece is testing what. For example, it isn't clear why `test-action` and `test-inputs` are separate workflow files. If someone were adding a new test, it isn't clear which file it should go in.

File: README.md [hunk 16]
```diff
@@ -1,14 +1,9 @@
 # cargo-semver-checks-action
 Lint your crate API changes for semver violations.
```
Comment (obi1kenobi) [inline README.md:2]:
```suggestion
Lint your crate API changes for semver violations.

*This is the work-in-progress v2 version of this action. Find the docs for the stable v1 action [here](https://github.com/obi1kenobi/cargo-semver-checks-action/tree/v1.4).*
```
Since the action isn't yet complete and you're still planning on adding major pieces of functionality.

**Review by obi1kenobi — COMMENTED** [review]
Minor polish items, I think this should be the last iteration before we merge this 🚀

File: .github/workflows/test-action.yml (hunk 15 above)
Comment (mgr0dzicki) [reply .github/workflows/test-action.yml:1]: Definitely a good idea, added a README. I also made two other changes that you suggested, please have a look now.

File: .github/workflows/setup-test-workspace/action.yml [hunk 17]
```diff
@@ -1,12 +1,17 @@
 name: Setup test workspace
+inputs:
+  ref-slice-ref:
+    description: 'Reference of the ref_slice to checkout'
+    required: false
+    default: ''
```
Comment (obi1kenobi) [inline .github/workflows/setup-test-workspace/action.yml:6]: I'm confused by this. If the input isn't set (because it isn't required) and the default is empty-string, what will `actions/checkout@v3` do? Will that be the right thing to do?

File: .github/workflows/ci.yml [hunk 18]
```diff
@@ -6,52 +6,20 @@ on:
     branches:
       - main
 
-env:
-  RUST_BACKTRACE: 1
-  CARGO_TERM_COLOR: always
-
```
Comment (obi1kenobi) [inline .github/workflows/ci.yml:12]: Any particular reason why we're removing these? They don't seem to be doing any harm and we want them around as a useful default in case other jobs get added in the future.

File: .github/workflows/setup-test-workspace/action.yml [hunk 19]
```diff
@@ -0,0 +1,24 @@
+name: Setup test workspace
+inputs:
+  ref-slice-ref:
+    description: 'Reference of the ref_slice to checkout'
```
Comment (obi1kenobi) [inline .github/workflows/setup-test-workspace/action.yml:4]: Consider being explicit about this being a *git* reference: a branch, tag, or commit hash. `Reference` is a generic term so this description doesn't communicate much beyond what the input name already got across.

**Review by obi1kenobi — COMMENTED** [review]
Just some lint-level nitpicks, otherwise looks ready to merge. I'm sure I only spotted them because I was looking at this in the PR diff view instead of an editor. consider doing a self-review of the PR to catch these yourself and reduce the number of back-and-forth iterations we need over tiny issues.

Once you've had a chance for that final tiny round of polish, just mark the PR as no longer a draft and I'd be happy to merge.

**Comment by mgr0dzicki** [comment]
> Just some lint-level nitpicks, otherwise looks ready to merge. I'm sure I only spotted them because I was looking at this in the PR diff view instead of an editor. consider doing a self-review of the PR to catch these yourself and reduce the number of back-and-forth iterations we need over tiny issues.

I have to admit, I was using the editor and not looked at the PR diff view, my mistake. Sorry for all these issues! I try to double-check everything, but this PR is quite huge and I'm looking at it for a week now, so finding such mistakes becomes increasingly difficult...

I'll correct what you found in a moment and then self-review everything as you suggested.

**Comment by obi1kenobi** [comment]
> Sorry for all these issues! I try to double-check everything, but this PR is quite huge and I'm looking at it for a week now, so finding such mistakes becomes increasingly difficult...

No worries and no need to apologize. Every experience is something we can learn from, and the top objective is the learning itself much more so than the artifact being produced in the PR itself.

I also feel the same way about the PR being huge and difficult to (re-)review over and over. In retrospect, both of us could probably have predicted this as a risk and structured it differently so we don't have to merge it all-or-nothing like this. We'll do better next time!

**Comment by mgr0dzicki** [comment]
Hope this is finally it... ;)

File: .github/workflows/setup-test-workspace/action.yml [hunk 20]
```diff
@@ -1,9 +1,8 @@
 name: Setup test workspace
 inputs:
   ref-slice-ref:
-    description: 'Reference of the ref_slice to checkout'
+    description: 'Git reference (branch name, tag or commit hash) of the ref_slice test fork to checkout'
     required: false
```
Comment (obi1kenobi) [inline .github/workflows/setup-test-workspace/action.yml:5]: Should this be `required: true`? Otherwise, what's the behavior if this input value isn't supplied?

Any time we make something optional, we're adding more modes of operation -- which means more things to test, and more things that can fail. Are we losing anything if we make this value required?

File: action.yml [hunk 21]
```diff
@@ -7,11 +7,9 @@ inputs:
   package:
     description: 'The package whose API to check for semver (in Package Id Specification format, see https://doc.rust-lang.org/cargo/reference/pkgid-spec.html for reference). If not set, all packages in the workspace are processed.'
     required: false
-    default: ''
   manifest-path:
     description: 'Path to Cargo.toml of crate or workspace to check.'
```
Comment (obi1kenobi) [inline action.yml:11]: The inputs here are generally great at documenting what the default behavior is when a `required: false` input isn't set. Let's document that behavior here as well.

**Review by obi1kenobi — APPROVED** [review]
It looks good enough to merge, let's do that and move forward. I'm highlighting two small additional issues that we should try to fix in a subsequent PR.

File: .github/workflows/setup-test-workspace/action.yml (hunk 20 above)
Comment (mgr0dzicki) [reply .github/workflows/setup-test-workspace/action.yml:5]: My aim was to make the default behavior the same as for the `checkout` action, but you're right - since we don't use it, it's better to make the input required then wonder what the default behavior exactly is and try documenting it here.

File: action.yml (hunk 21 above)
Comment (mgr0dzicki) [reply action.yml:11]: I've done this and the above one in https://github.com/obi1kenobi/cargo-semver-checks-action/pull/25
